resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/FailureWithinTimeIntervalTracker.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.yarn
+
+import scala.collection.mutable
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Clock, SystemClock}
+
+private[spark] class FailureWithinTimeIntervalTracker(sparkConf: SparkConf) extends Logging {
+
+  private var clock: Clock = new SystemClock
+
+  private val executorFailuresValidityInterval =
+    sparkConf.get(config.EXECUTOR_ATTEMPT_FAILURE_VALIDITY_INTERVAL_MS).getOrElse(-1L)
+
+  // Queue to store the timestamp of failed executors for each host
+  private val failedExecutorsTimeStampsPerHost = mutable.Map[String, mutable.Queue[Long]]()
+
+  private val failedExecutorsTimeStamps = new mutable.Queue[Long]()
+
+  private def getNumFailuresWithinValidityInterval(
+      failedExecutorsTimeStampsForHost: mutable.Queue[Long],
+      endTime: Long): Int = {
+    while (executorFailuresValidityInterval > 0
+      && failedExecutorsTimeStampsForHost.nonEmpty
+      && failedExecutorsTimeStampsForHost.head < endTime - executorFailuresValidityInterval) {
+      failedExecutorsTimeStampsForHost.dequeue()
```

Comment:
It's counter-intuitive that this `get*` method mutates state. If I called 

    getNumFailuresWithinValidityInterval(foo, 0)
    getNumFailuresWithinValidityInterval(foo, 10)
    getNumFailuresWithinValidityInterval(foo, 0)

The last call can return something different from the first because all the failures that weren't within `10 - executorFailuresValidityInterval` will have been dropped.

Reply:
Ok, I will take your recommendation and drop endTime with renaming the method to  getRecentFailureCount.